// angular-auth-oidc-client/esm2015/lib/services/oidc.security.config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { of, ReplaySubject } from 'rxjs';
import { catchError, switchMap } from 'rxjs/operators';
/**
 * @record
 */
export function ConfigResult() { }
if (false) {
    /** @type {?} */
    ConfigResult.prototype.authWellknownEndpoints;
    /** @type {?} */
    ConfigResult.prototype.customConfig;
}
export class OidcConfigService {
    /**
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
        this.configurationLoadedInternal = new ReplaySubject(1);
    }
    /**
     * @return {?}
     */
    get onConfigurationLoaded() {
        return this.configurationLoadedInternal.asObservable();
    }
    /**
     * @param {?} configUrl
     * @return {?}
     */
    load(configUrl) {
        return this.httpClient
            .get(configUrl)
            .pipe(switchMap((/**
         * @param {?} clientConfiguration
         * @return {?}
         */
        clientConfiguration => {
            return this.loadUsingConfiguration(clientConfiguration);
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.error(`OidcConfigService 'load' threw an error on calling ${configUrl}`, error);
            this.configurationLoadedInternal.next(undefined);
            return of(false);
        })))
            .toPromise();
    }
    /**
     * @param {?} stsServer
     * @return {?}
     */
    load_using_stsServer(stsServer) {
        return this.loadUsingConfiguration({ stsServer }).toPromise();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    load_using_custom_stsServer(url) {
        return this.httpClient
            .get(url)
            .pipe(switchMap((/**
         * @param {?} wellKnownEndpoints
         * @return {?}
         */
        wellKnownEndpoints => {
            this.configurationLoadedInternal.next({
                authWellknownEndpoints: wellKnownEndpoints,
                customConfig: { stsServer: url },
            });
            return of(true);
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.error(`OidcConfigService 'load_using_custom_stsServer' threw an error on calling ${url}`, error);
            this.configurationLoadedInternal.next(undefined);
            return of(false);
        })))
            .toPromise();
    }
    /**
     * @private
     * @param {?} clientConfig
     * @return {?}
     */
    loadUsingConfiguration(clientConfig) {
        if (!clientConfig.stsServer) {
            console.error(`Property 'stsServer' is not present of passed config ${JSON.stringify(clientConfig)}`, clientConfig);
            throw new Error(`Property 'stsServer' is not present of passed config ${JSON.stringify(clientConfig)}`);
        }
        /** @type {?} */
        const url = `${clientConfig.stsServer}/.well-known/openid-configuration`;
        return this.httpClient.get(url).pipe(switchMap((/**
         * @param {?} wellKnownEndpoints
         * @return {?}
         */
        wellKnownEndpoints => {
            this.configurationLoadedInternal.next({
                authWellknownEndpoints: wellKnownEndpoints,
                customConfig: clientConfig,
            });
            return of(true);
        })), catchError((/**
         * @param {?} error
         * @return {?}
         */
        error => {
            console.error(`OidcConfigService 'load_using_stsServer' threw an error on calling ${url}`, error);
            this.configurationLoadedInternal.next(undefined);
            return of(false);
        })));
    }
}
OidcConfigService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
OidcConfigService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    OidcConfigService.prototype.configurationLoadedInternal;
    /**
     * @type {?}
     * @private
     */
    OidcConfigService.prototype.httpClient;
}
